Feature/iris/iris tutor suggestions communication

alexjoham opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [ ] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [ ] I **strictly** followed the principle of **data economy** for all database calls.
- [ ] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [ ] I documented the Java code using JavaDoc style.


#### Client
- [ ] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [ ] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [ ] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [ ] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.


#### Changes affecting Programming Exercises
- [ ] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **LocalVC** and **Jenkins**.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->


### Description
<!-- Describe your changes in detail -->


### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis
2. Navigate to Course Administration
3. ...

#### Exam Mode Testing
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that the UI of the programming exercise in the exam mode stays unchanged. You can use the [exam mode documentation](https://docs.artemis.cit.tum.de/user/exam_mode/) as reference.
4. ...

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
<!-- See [Large Course Setup](https://github.com/ls1intum/Artemis/tree/develop/supporting_scripts/course-scripts/quick-course-setup) for the automation script that handles large course setup. -->
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files modified in this PR here. You can use `supporting_script/code-coverage/generate_code_cov_table/generate_code_cov_table.py` to automatically generate the coverage table from the corresponding artefacts of your branch (follow the ReadMe for setup details). -->
<!-- Alternatively you can execute the tests locally (see build.gradle and package.json) or look into the corresponding artefacts. -->
<!-- The line coverage must be above 90% for changes files, and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

alexjoham commented on src/main/webapp/app/iris/overview/services/iris-chat-http.service.ts:
```diff
@@ -70,6 +70,21 @@ export class IrisChatHttpService {
         );
     }
 
+    /**
+     * Creates a new tutor suggestion message in a session
+     * @param sessionId of the session
+     * @param message to be created
+     */
+    createTutorSuggestion(sessionId: number, message: IrisTutorSuggestionRequestMessage): Response<IrisTutorSuggestionRequestMessage> {
+        return this.httpClient.post<IrisTutorSuggestionRequestMessage>(
+            `${this.apiPrefix}/sessions/${sessionId}/messages`,
+            Object.assign({}, message, {
+                sentAt: convertDateFromClient(message.sentAt),
+            }),
+            { observe: 'response' },
+        );
+    }
```
There is no messageDifferentiator in IrisTutorSuggestionRequestMessage therefore this is not needed

sebastianloose commented on src/main/webapp/app/communication/course-conversations/tutor-suggestion/tutor-suggestion.component.ts:
```diff
@@ -0,0 +1,149 @@
+                catchError(() => of([])),
+            )
+            .subscribe((messages) => {
+                const lastMessage = messages[messages.length - 1];
```
Better check the length of the messages array first. Otherwise you might raise an IndexOutOfBounds if the array is empty

alexjoham replied:
Typescript returns undefined when accessing an empty string with a negative index, so in my case nothing happens, as i am working with lastMessage?.sender afterwards.

sebastianloose commented on src/main/webapp/app/communication/course-conversations/tutor-suggestion/tutor-suggestion.component.ts:
```diff
@@ -0,0 +1,149 @@
+                    if (post) {
+                        this.tutorSuggestionSubscription = this.chatService
+                            .requestTutorSuggestion()
+                            .pipe(catchError(() => of(undefined)))
```
I think it would be better to display the error or log it rather than just ignoring it

alexjoham replied:
I will into that

alexjoham replied:
Done

sebastianloose commented on src/main/webapp/app/communication/course-conversations/tutor-suggestion/tutor-suggestion.component.ts:
```diff
@@ -0,0 +1,149 @@
+        this.errorSubscription = this.chatService.currentError().subscribe((error) => (this.error = error));
+    }
+
+    protected readonly IrisLogoSize = IrisLogoSize;
+    protected readonly IrisTextMessageContent = IrisTextMessageContent;
```
Move this read only attributes to the top of the component. It is clearer if all attributes are defined there

isabellagessl commented on src/main/webapp/app/communication/course-conversations-components/layout/conversation-thread-sidebar/conversation-thread-sidebar.component.html:
```diff
@@ -62,6 +62,11 @@ <h5 class="mb-0 fw-medium" jhiTranslate="artemisApp.conversationsLayout.threadSi
                     </div>
                 }
             </div>
+            @if (isAtLeastTutor && !post?.resolved) {
```
You cpould move this checks directly into the tutor suggestion component as it should only be used/ seen when the user is a tutor and the post is not resolved. Not that someone uses the component but forgets to add the extra checks (is user a tutor & post not resolved). you could also remove the <div> around the tutor suggestion here because you already have a <div> in the component itself.

Anishyou commented:
![image](https://github.com/user-attachments/assets/0c0fa768-6848-44e4-a1a8-682a41c5a0ee)

Iris suggestion needs to be aligned with the dark mode

![image](https://github.com/user-attachments/assets/a0191842-d9a3-4e06-8842-f8566ddec1db)
When clicked on the selection box, it shows nothing

alexjoham commented:
I added a darkmode. Regarding the bug with the selected variant. I had it too locally, but also for other Iris settings. I deleted the settings in the database and restarted the Server and not it is working again displaying default as a variant. I am not sure what is destroying the database entries, but I reused exisiting settings to build those settings, so if this problem still exists we should think about creating an issue for that

Anishyou commented:
Tested locally- Works fine. On student perspective tutor suggestion pipeline is not shown.
For tutor perspective and users above tutors like instructor and admin , tutor suggestion pipeline works.
👍

Anishyou reviewed (approved):
Locally tested. Here are some screenshots 👍
![image](https://github.com/user-attachments/assets/d87f210a-24cc-41e3-a3cd-15c0fddb0d55)
![image](https://github.com/user-attachments/assets/1b8268bc-b709-4d16-a864-8a1a2b5004a9)
![image](https://github.com/user-attachments/assets/bea0c8a0-c312-45d7-84b8-5012ee05b8d0)

krusche commented on src/main/resources/config/liquibase/changelog/20250326145226_changelog.xml:
```diff
@@ -0,0 +1,32 @@
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.31.xsd"
+        objectQuotingStrategy="QUOTE_ONLY_RESERVED_WORDS">
```
never saw this, do we really need it?

alexjoham replied:
No we don't need this, I removed it

krusche commented on src/main/resources/config/liquibase/changelog/20250326145226_changelog.xml:
```diff
@@ -0,0 +1,32 @@
+                                 constraintName="FK_IRIS_SESSION_ON_POST"
+                                 referencedColumnNames="id"
+                                 referencedTableName="post"
+                                 onDelete="CASCADE"
```
does this mean the post is deleted when the iris session is deleted or the other one round? we need to be careful

alexjoham replied:
This means the iris session is deleted if a post is deleted

krusche commented on src/main/java/de/tum/cit/aet/artemis/iris/domain/settings/IrisTutorSuggestionSubSettings.java:
```diff
@@ -0,0 +1,16 @@
+@Entity
+@DiscriminatorValue("TUTOR_SUGGESTION")
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class IrisTutorSuggestionSubSettings extends IrisSubSettings {
```
Do we really need so many iris settings? It becomes more and more complicated. I do not really think that instructors want to specify so many details

alexjoham replied:
We can discuss how we could connect the feature to other settings of Iris, but being able to disable it should be possible in my opinion

krusche replied:
The problem is that we have so many different settings that can be enabled and disabled. I don't think that instructors need so many options. This simply leads to the fact, that the feature is not used at all, because it is disabled by default, no instructor knows about the feature and even if they would know, they would not find how to enable it.

We have the lecture "ingestion" feature for more than 6 months now, however no one has used it so far :-( 

Also this means that everytime, we need to load settings if something is enabled or not in the client user interface. This also leads to issues.

krusche replied:
I believe there should be a simple switch for instructors to either enable IRIS completely or disable it completely. We could have an intermediate state where the "most important" and "most stable" features are active, but we really do not need so many settings. Think about other tools that you use. Are you enabled or disabling all features with lots of different settings or do you trust the software provider that it works? Using the features is still voluntary (and requries opt-in), no one is forced.

krusche commented on src/main/java/de/tum/cit/aet/artemis/iris/domain/settings/IrisCourseSettings.java:
```diff
@@ -52,6 +52,10 @@ public class IrisCourseSettings extends IrisSettings {
     @JoinColumn(name = "iris_competency_generation_settings_id")
     private IrisCompetencyGenerationSubSettings irisCompetencyGenerationSettings;
 
+    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
```
The many eager fetches here can become a problem with Hibernate. We definitely need to change this in the future, this has become way too complex when you think about that a few boolean values are stored.

krusche replied:
@Hialus I suggest we talk about this asap

Anishyou commented:
![image](https://github.com/user-attachments/assets/2517a400-59b7-4c82-a58b-ef3d79a76a10)
There is an issue with the UI

Anishyou commented:
![image](https://github.com/user-attachments/assets/0630da89-8ff9-45e3-8f95-8197d53de6a2)
I have a question here- I posted this question in lecture communication, shouldn't it reply about that specific lecture?

alexjoham commented on src/main/java/de/tum/cit/aet/artemis/iris/domain/session/IrisTutorSuggestionSession.java:
```diff
@@ -0,0 +1,48 @@
+
+    @ManyToOne
+    @JsonIgnore
+    private Post post;
```
I need most of the information the Post contains. I could only save the postId and fetch the Post via the repository, as every IrisSession saves its reference that way, I wanted to stay consistent with the implementation.

krusche replied:
I would argue to change this in other places as well, otherwise we might get severe performance issues. Do you really always need the post information when you load the session? Or is there one specific case where you need this?

alexjoham replied:
I updated this to the postId and no the the post is fetched with the repository only when its needed

krusche commented on src/main/resources/config/liquibase/changelog/20250326145226_changelog.xml:
```diff
@@ -0,0 +1,29 @@
+                                 referencedTableName="iris_sub_settings"
+                                 onDelete="CASCADE"
+        />
+    </changeSet>
```
can you add the rollback sql script here?

Anishyou reviewed (changes requested):
![image](https://github.com/user-attachments/assets/c97b8f6a-46b0-458a-a949-884e1bfdcf02)
It is ok for iris setting in course, but for iris settings in server administration, this is still the issue

Anishyou commented:
Other than the above UI bug,  I tested manually, and it works as expected.

alexjoham commented:
> ![image](https://private-user-images.githubusercontent.com/123313052/440752794-c97b8f6a-46b0-458a-a949-884e1bfdcf02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1MzEyOTYsIm5iZiI6MTc0NjUzMDk5NiwicGF0aCI6Ii8xMjMzMTMwNTIvNDQwNzUyNzk0LWM5N2I4ZjZhLTQ2YjAtNDU4YS1hOTQ5LTg4NGUxYmZkY2YwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQxMTI5NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMmIzM2RmNzU2Mzk2YThmZmVkZGQ0ZjJkN2M5YTc0MzA3OTk1N2JlZjNjNjBjZmQyZmRmNWIwZTA2OWNjZWFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HF-xGB62Necx13Yjc6aemplXDgwzfoUljXh-5pvvivs) It is ok for iris setting in course, but for iris settings in server administration, this is still the issue

This was updated with one of the last commits. Unfortunately, you must delete the old Iris setting in the database and restart the backend to have it work normally.

Anishyou commented:
> > ![image](https://private-user-images.githubusercontent.com/123313052/440752794-c97b8f6a-46b0-458a-a949-884e1bfdcf02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1MzEyOTYsIm5iZiI6MTc0NjUzMDk5NiwicGF0aCI6Ii8xMjMzMTMwNTIvNDQwNzUyNzk0LWM5N2I4ZjZhLTQ2YjAtNDU4YS1hOTQ5LTg4NGUxYmZkY2YwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwNlQxMTI5NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMmIzM2RmNzU2Mzk2YThmZmVkZGQ0ZjJkN2M5YTc0MzA3OTk1N2JlZjNjNjBjZmQyZmRmNWIwZTA2OWNjZWFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HF-xGB62Necx13Yjc6aemplXDgwzfoUljXh-5pvvivs) It is ok for iris setting in course, but for iris settings in server administration, this is still the issue
> 
> @Anishyou This was updated with one of the last commits. Unfortunately, you must delete the old Iris setting in the database and restart the backend to have it work normally.

Ok, It works now, Thanks

Anishyou reviewed (approved):
Locally tested. Works as mentioned

sebastianloose reviewed (dismissed):
Code looks good to me! Tested locally

isabellagessl reviewed (dismissed):
Code lgtm approve.

isabellagessl reviewed (approved):
Code looks good to me!

Anishyou reviewed (approved):
Tested locally works as expected. re-approve

sebastianloose reviewed (approved):
Reapprove! Code looks still good

armanhabibiTUM reviewed (approved):
Tested on test server 3. Works as expected.

<img width="1280" alt="Bildschirmfoto 2025-05-19 um 11 41 40 AM" src="https://github.com/user-attachments/assets/ca24a55b-51b5-43fb-b197-d1d631db2078" />
